### PR TITLE
Rewrite of SSE2 `update_cdf`

### PR DESCRIPTION
**Pull request by dodomorandi**
A WIP rewrite of the SSE2 version of `update_cdf`, issue https://github.com/xiph/rav1e/issues/1720.

Current resulting assembly is pretty similar to the one written by hand in `dav1d`. I would like to see if I can do slightly better, and I would like to write a small documentation of what is happening.

### CODE REVIEW

**Comment by dodomorandi** [comment]
I think I have an idea about the problem. Unfortunately the `update_cdf` test is passing flawlessly, I probably need to write a specific regression test for this.

**Comment by dodomorandi** [comment]
This is not the best solution: if the content of the auxiliary `k` variable is `-1` when `index >= val` instead of 0, it should be possible to replicate the behavior of `dav1d` and remove a SIMD instruction. However, I am not getting how to do that in a cheap way.

**Comment by tdaede** [comment]
I'll take a look to see if that last instructions can be dropped, but the rest looks ok.

**Comment by lu-zero** [comment]
Could you please rebase this set?

**Comment by tdaede** [comment]
I'll also take a look in the next day to see if there's a way to drop the last insn.

**Comment by dodomorandi** [comment]
I tried to find a cheaper solution, but without success. I am removing the `WIP` tag, but @tdaede if you find I way to optimize the code, suggestions are welcome.

**Comment by tdaede** [comment]
Yeah, I think this is fine. If I come up with something better in the meantime I'll open that as a PR.

**Comment by tdaede** [comment]
You'll need to squash the commits though. (or I can do it for you)

**Comment by dodomorandi** [comment]
:+1: I was not sure if you would rather do a squash-merge.

File: src/asm/x86/ec.rs [hunk 1]
```diff
@@ -24,29 +24,66 @@ pub fn update_cdf(cdf: &mut [u16], val: u32) {
 #[target_feature(enable = "sse2")]
 unsafe fn update_cdf_4_sse2(cdf: &mut [u16], val: u32) {
   let nsymbs = 4;
-  let rate = 3 + (nsymbs >> 1).min(2) + (cdf[nsymbs] >> 4) as usize;
+  let rate = 5 + (cdf[nsymbs] >> 4) as usize;
   cdf[nsymbs] += 1 - (cdf[nsymbs] >> 5);
```
Comment (barrbrain) [inline src/asm/x86/ec.rs]: To further align with dav1d assembly:
```suggestion
  cdf[nsymbs] += (cdf[nsymbs] < 32) as u16;
```

Comment (dodomorandi) [reply src/asm/x86/ec.rs]: Good catch, I overlooked that part. And honestly I overestimated the capabilities of the optimizer -- your solution produces a cleaner asm 👍

**Comment by lu-zero** [comment]
Looks fine to me, I'd change the commit title to "Improve the sse2 update_cdf" though

**Comment by lu-zero** [comment]
Let's land it then

**Pull request merged**